lib/utils.js

```diff
@@ -132,7 +132,9 @@ exports.compact = function (obj, references) {
         var compacted = [];
 
         for (var i = 0; i < obj.length; ++i) {
-            if (typeof obj[i] !== 'undefined') {
+            if (Array.isArray(obj[i]) || Object.prototype.toString.call(obj[i]) === '[object Object]') {
```

Comment:
i wonder tho if this might still be a simpler check to do `obj[i] && typeof obj[i] === 'object'` which would catch all non-function primitives?